### DISCUSSION TITLE
Request for a architecture overview

### DISCUSSION THREAD

**Discussion by code-withAshish** [pocketbase/pocketbase#2448] [Q&A]
I am a college student and for my final year project i want to make a rewrite of pocketbase itself in rust so it would help me learning rust and gets my project work done.
So if someone could explain me the architecture of pocketbase and how it works that would be very helpful, as I am not very much familiar with golang.
Thank you

**Comment by ganigeorgiev** [comment]
You can go through the [Basics docs section](https://pocketbase.io/docs/) to get a general overview what PocketBase offer and how it manage its data.

> So if someone could explain me the architecture of pocketbase and how it works that would be very helpful

This is difficult to answer as the question is too broad and it'll require spending too much time to go through everything (_and I doubt that it will be helpful_). If you have a specific question and want to know how a certain part works, feel free to ask and I'll make an attempt to explain it, but please try to provide as much details as possible with your questions.

My general recommendation would be to try to explore the code even if you are not familiar with Go (for example you can start with the [core package](https://github.com/pocketbase/pocketbase/tree/master/core) and from there move on to the other packages; _also keep in mind that some of the internals, specifically the hooks, may change in the near future_).

You may also find helpful skimming through the ["A Tour of Go"](https://go.dev/tour/welcome/1) to get a general idea of the syntax and the available primitives.

**Reply by code-withAshish** [reply]
hey thank you so much for the reply
and if i had to ask something very specific about what i want to know "Coukd you please explain how do you manage th sqlite3 db"? like do you implement concurrency or sharding features or it's just normal sqlite instance running and how do you implement real time subscriptions i am assuming web-sockets but will be helpful if you could guide me to correct path or correct me if i am wrong.
Thank you

**Reply by ganigeorgiev** [reply]
> Coukd you please explain how do you manage th sqlite3 db

We embed SQLite either with [mattn/go-sqlite3](https://github.com/mattn/go-sqlite3) using CGO (it comes with the [SQLite amalgamation](https://www.sqlite.org/amalgamation.html)) or use the Go transpiled port [cznic/sqlite](https://gitlab.com/cznic/sqlite), depending whether `CGO_ENABLED=0/1` is used.

The SQLite database is opened in [WAL mode](https://www.sqlite.org/wal.html), which allow us to have multiple readers and a single writer (writes queue up) to execute concurrently.

---

> like do you implement concurrency or sharding features 

In additional to the builtin [`busy_timeout` PRAGMA](https://www.sqlite.org/pragma.html#pragma_busy_timeout) we also have auto fail retry for the read and write queries on application level as fallback for some edge cases to minimize as much as possible the eventual [SQLITE_BUSY](https://www.sqlite.org/rescode.html#busy) errors (_for example you can check [base_retry.go](https://github.com/pocketbase/pocketbase/blob/master/daos/base_retry.go)_).
PocketBase can scale only on a single server and there is no sharding out of the box.

---

> how do you implement real time subscriptions

[Server-sent events (SSE)](https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events) are used for the realtime subscriptions. The main difference from WebSocket is that SSE is unidirectional, aka. only the server can send messages over the persistent connection. In general both approaches are valid. I personally prefer working with SSE over WebSockets whenever possible because it is easier to handle as there is no need for the "hacky" HTTP(s)->WS(s) protocol upgrade (SSE is a regular long-lived HTTP connection) and it works well when combined with other REST/RPC apis since you don't have to duplicate for example the business logic for submitting a record as you can use regular REST/RPC calls to submit messages client->server.

We use a simple fanout-like pattern (_practically iterating over each subscriber/client_) for broadcasting a single record event concurrently to all connected clients. The only slightly more complicated part probably is the access checks which are based on the [collection List and View API rules](https://pocketbase.io/docs/api-rules-and-filters/) (_rust has macros so you [should be able create DSLs](https://doc.rust-lang.org/rust-by-example/macros/dsl.html) to accomplish something similar, use some external library or some other approach_).

**Reply by code-withAshish** [reply]
hey i was able to do a lot of things with your explanation and i am thankful for that, but one thing that i was thinking is, sending direct sql queries from the client (pocketbase js sdk) to the pocketbase instance is safe?

or should i end json structure and build the queries myself on server itself?

how does pocketbase handle this currently i am doing the raw sql approach, but i was wondering if it's safe or not?

Thank you

**Reply by ganigeorgiev** [reply]
> sending direct sql queries from the client (pocketbase js sdk) to the pocketbase instance is safe?

We don't do this, neither we have support for it (_or at least not out of the box_). For the available web APIs please refer to https://pocketbase.io/docs/api-records/.

**Reply by code-withAshish** [reply]
so how exactly does pocketbase does this? 
i  mean in simple terms what does pocketbase clients send to pocketbase instances to do CRUD? 
and how does pockebase handles this

**Comment by code-withAshish** [comment]
This was pretty helpful insight into the pocketbase, thank you for this
but for the time being i would request to not close this thread as i would like to ask more questions on the same.
Thank you for your time @ganigeorgiev
